File: SinaWeibo/Views/AvatarView.m
//
//  AvatarView.m
//  SinaWeibo
//
//  Created by cxjwin on 6/13/14.
//  Copyright (c) 2014 cxjwin. All rights reserved.
//


#import <SDWebImage/SDWebImageManager.h>
#import "AvatarView.h"

@implementation AvatarView

- (instancetype)initWithCoder:(NSCoder *)coder {
	self = [super initWithCoder:coder];
	if (self) {
		//
	}

	return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
	self = [super initWithFrame:frame];
	if (self) {
		// Initialization code
	}

	return self;
}

- (void)setURLString:(NSString *)URLString {
	if (_URLString != URLString) {
		_URLString = [URLString copy];
	}
	
	NSURL *avatarUrl = nil;
	if (_URLString) {
		avatarUrl = [NSURL URLWithString:_URLString];
	}
	
	if (avatarUrl) {
		SDWebImageManager *manager = [SDWebImageManager sharedManager];
		UIImage *image = [manager.imageCache imageFromMemoryCacheForKey:[avatarUrl absoluteString]];
		if (image) {
			self.image = image;
		} else {
			self.image = nil;
			
			__weak typeof(self) weakSelf = self;
			void (^completed)(UIImage *, NSError *, SDImageCacheType, BOOL) = ^(UIImage *image, NSError *error, SDImageCacheType cacheType, BOOL finished) {
				if (image) {
					NSAssert([NSThread isMainThread], @"is main thread");
					dispatch_async(dispatch_get_main_queue(), ^{
						weakSelf.image = image;
					});
				}
			};
			
			[manager downloadWithURL:avatarUrl options:SDWebImageCacheMemoryOnly progress:nil completed:completed];
		}
	}
}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect
{
    // Drawing code
}
*/

@end

File: SinaWeibo/Views/AvatarView.swift
//
//  AvatarView.swift
//  SinaWeibo
//
//  Created by cxjwin on 6/17/14.
//  Copyright (c) 2014 cxjwin. All rights reserved.
//

import UIKit

@objc(AvatarView)
@IBDesignable
class AvatarView: UIImageView {

	@IBInspectable var URLString: String? {
	didSet {
		if let _URLString = URLString {
			var avatarUrl = NSURL(string: _URLString);
			
			let manager = SDWebImageManager.sharedManager()
			var cacheimage: UIImage? = manager.imageCache.imageFromDiskCacheForKey(avatarUrl.absoluteString)
			
			if let _image = cacheimage {
				self.image = _image
			} else {
				self.image = nil
				
				var completed: (image: UIImage?, error: NSError?, cacheType: SDImageCacheType, finished: Bool) -> Void = {
					[weak self] (image: UIImage?, error: NSError?, cacheType: SDImageCacheType, finished: Bool) in
					if (image != nil) {
                        self?.image = image
					}
				}
				
				manager.downloadWithURL(avatarUrl, options: SDWebImageOptions.CacheMemoryOnly, progress: nil, completed: completed)
			}
		}
	}
	}
	
    override init(frame: CGRect) {
        super.init(frame: frame)
        // Initialization code
    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    /*
    // Only override drawRect: if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func drawRect(rect: CGRect)
    {
        // Drawing code
    }
    */

}
